File: app/models/businesses.py
from .db import db


class Business(db.Model):
  __tablename__ = 'businesses'

  id = db.Column(db.Integer, primary_key = True)
  user_id = db.Column(db.Integer, db.ForeignKey("users.id"))
  name = db.Column(db.String(100), nullable = False, unique = True)
  description = db.Column(db.Text, nullable = False)
  lat = db.Column(db.String(20), nullable=False)
  lng = db.Column(db.String(20), nullable=False)
  address = db.Column(db.String(100), nullable=True)
  city = db.Column(db.String(50), nullable=False)
  state = db.Column(db.String(15), nullable=False)
  zipcode = db.Column(db.String(10), nullable=False)
  website = db.Column(db.String, nullable=True)
  contact = db.Column(db.String, nullable=False)
  imgURL = db.Column(db.String, nullable=False)

  types = db.relationship('BusinessType', back_populates='business', cascade="all, delete-orphan") # lazy='joined')
  judgements = db.relationship('Judgement', back_populates='business', cascade="all, delete-orphan") #  lazy='joined')
  user = db.relationship('User', back_populates='businesses',) #lazy='joined')
  reviews = db.relationship('Review', back_populates='business', cascade="all, delete-orphan" ) #lazy='joined')

  def to_dict(self):
    return {
      "id": self.id,
      "user_id": self.user_id,
      "name": self.name,
      "description": self.description,
      "lat": self.lat,
      "lng": self.lng,
      "address": self.address,
      "city": self.city,
      "state": self.state,
      "zipcode": self.zipcode,
      "website": self.website,
      "contact": self.contact,
      "imgURL": self.imgURL,
      "reviews": [review.to_dict() for review in self.reviews],
      "types": [type.to_dict() for type in self.types]
    }

File: react-app/src/services/businesses.js
export const fetchBusinesses = async () => {
  const response = await fetch('/api/business/', {
    headers: {
      'Content-Type': 'application/json'
    },
  });
  const business = await response.json();
  return business.businesses
}

export const getBusiness = async (id) => {
  const response = await fetch(`/api/business/${id}`, {
    headers: {
      'Content-Type': 'application/json'
    },
  });
  const business = await response.json();
  return business
}

export const addBusinessType = async (business_id, typeId) => {
  const response = await fetch(`/api/business/${business_id}`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      business_id,
      typeId
    }),
  });
  return await response.json();
}

export const newBusiness = async (user_id, name, description,
  lat, lng, address, city, state, zipcode, website, contact, imgURL) => {
  const response = await fetch(`/api/business/`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      user_id,
      name,
      description,
      lat,
      lng,
      address,
      city,
      state,
      zipcode,
      website,
      contact,
      imgURL,
    }),
  });
  return await response.json();
}

export const newBusinessType = async (title) => {
  const response = await fetch(`/api/business/types`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ title } ),
  });
  const newType = await response.json();
  console.log(newType, "newType")
  return newType
}

export const fetchTypes = async () => {
  const response = await fetch('/api/business/types', {
    headers: {
      'Content-Type': 'application/json'
    },
  });
  const type = await response.json();
  return type.types
}

export const sendUpdatedBusiness = async (id, name, lat, lng, address, city, state, zipcode, website, contact) => {
  const response = await fetch(`api/business/${id}`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      id,
      name,
      lat,
      lng,
      address,
      city,
      state,
      zipcode,
      website,
      contact
    })
  })
    return await response.json();
}

export const deleteBusiness = async (id) => {
  const response = await fetch(`api/business/${id}`, {
    method: "DELETE",
    headers: {
      "Content-Type": "application/json",
    },
  })
    return await response.json();
}
